lxd/device/unix_hotplug.go

```diff
@@ -0,0 +1,215 @@
+// +build linux,cgo
+
+package device
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/farjump/go-libudev"
+
+	deviceConfig "github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance/instancetype"
+	"github.com/lxc/lxd/shared"
+)
+
+// unixHotplugIsOurDevice indicates whether the unixHotplug device event qualifies as part of our device.
+// This function is not defined against the unixHotplug struct type so that it can be used in event
+// callbacks without needing to keep a reference to the unixHotplug device struct.
+func unixHotplugIsOurDevice(config deviceConfig.Device, unixHotplug *UnixHotplugEvent) bool {
+	// Check if event matches criteria for this device, if not return.
+	if (config["vendorid"] != "" && config["vendorid"] != unixHotplug.Vendor) || (config["productid"] != "" && config["productid"] != unixHotplug.Product) {
+		return false
+	}
+
+	return true
+}
```

Comment:
Maybe this is intentional but if: `config["vendorid"] == ""` and `config["productid"] == ""` then this would return true which seems weird to me. Is this on purpose. If not I'd expect this to be:
```go
func unixHotplugIsOurDevice(config deviceConfig.Device, unixHotplug *UnixHotplugEvent) bool {
	// Check if event matches criteria for this device, if not return.
	if config["vendorid"] == "" && config["productid"] == "" {
		return false
	}

	if config["vendorid"] == unixHotplug.Vendor && config["productid"] == unixHotPlug.Product {
		return true
	}

	return false
}
```

Reply:
I assume you know that vendorid and productid can't be empty but it might still be worth it to verify _shrug_.

Reply:
I suspect the matching logic comes from our USB logic where productid and vendorid are both optional, having neither set matches everything.

For unix-hotplug, we currently expect at least one of productid or vendorid to be set as we don't want to allow passing every single hotplug device into the instance.

Reply:
@brauner 
When we used udevadm to look at the block subsystem kernel uevents, they didn't have product/vendor identifying info, unlike the USB subsystem kernel uevent. For ex., 

USB: 

KERNEL[19272.213207] add      /devices/pci0000:00/0000:00:14.0/usb3/3-1 (usb)
ACTION=add
BUSNUM=003
DEVNAME=/dev/bus/usb/003/052
DEVNUM=052
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb3/3-1
DEVTYPE=usb_device
MAJOR=189
MINOR=307
PRODUCT=8633/800a/100
SEQNUM=17227
SUBSYSTEM=usb
TYPE=0/0/0

block:

KERNEL[19273.248237] add      /devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/host6/target6:0:0/6:0:0:0/block/sdc (block)
ACTION=add
DEVNAME=/dev/sdc
DEVPATH=/devices/pci0000:00/0000:00:14.0/usb3/3-1/3-1:1.0/host6/target6:0:0/6:0:0:0/block/sdc
DEVTYPE=disk
MAJOR=8
MINOR=32
SEQNUM=17241
SUBSYSTEM=block

Reply:
Yeah, the `block` subsystem does not have vendors or product ids iiuc (/cc @stgraber).

Reply:
`udev` will ad product and vendor by calling out to ata_id and scsci_id which are both part of systemd-udevd and do horrible things such as ATA over SCSCI for NVME and that udev adds this info to the libudev event. _ugh_

Reply:
@brauner 
Is there a different way to identify devices via uevents that you'd suggest or are udev uevents fine for now? I couldn't quite parse what you meant by fields available for Driver core.